hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -838,37 +864,65 @@ bool MDMParser::powerOff(void)
     LOCK();
     bool ok = false;
     bool continue_cancel = false;
+    bool check_ri = false;
     if (_init && _pwr) {
         MDM_INFO("\r\n[ Modem::powerOff ] = = = = = = = = = = = = = =");
         if (_cancel_all_operations) {
             continue_cancel = true;
             resume(); // make sure we can use the AT parser
         }
         for (int i=0; i<3; i++) { // try 3 times
-            sendFormated("AT+CPWROFF\r\n");
-            int ret = waitFinalResp(NULL,NULL,40*1000);
-            if (RESP_OK == ret) {
-                _pwr = false;
-                // todo - add if these are automatically done on power down
-                //_activated = false;
-                //_attached = false;
-                // Give the modem some time to switch off cleanly
-                HAL_Delay_Milliseconds(1000);
-                ok = true;
+            if(!_atOk()) {
+                MDM_INFO("\r\n[ Modem::powerOff ] AT Failure, trying PWR_UC...");
+                HAL_GPIO_Write(PWR_UC, 0);
+                // >1.5 seconds on SARA R410M
+                // >1 second on SARA U2
+                // plus a little extra for good luck
+                HAL_Delay_Milliseconds(1600);
+                HAL_GPIO_Write(PWR_UC, 1);
+                if (_dev.dev == DEV_SARA_R410) {
+                    check_ri = true;
+                }
                 break;
+            } else {
+                sendFormated("AT+CPWROFF\r\n");
+                int ret = waitFinalResp(nullptr, nullptr, CPWROFF_TIMEOUT);
+                if (RESP_OK == ret) {
+                    _pwr = false;
+                    // todo - add if these are automatically done on power down
+                    //_activated = false;
+                    //_attached = false;
+                    ok = true;
+                    check_ri = true;
+                    break;
+                }
+                else if (RESP_ABORTED == ret) {
+                    MDM_INFO("\r\n[ Modem::powerOff ] found ABORTED, retrying...");
+                }
+                else {
+                    MDM_INFO("\r\n[ Modem::powerOff ] timeout, retrying...");
+                }
             }
-            else if (RESP_ABORTED == ret) {
-                MDM_INFO("\r\n[ Modem::powerOff ] found ABORTED, retrying...");
-            }
-            else {
-                MDM_INFO("\r\n[ Modem::powerOff ] timeout, retrying...");
-            }
         }
     }
 
+    if (check_ri) {
+        system_tick_t t0 = HAL_Timer_Get_Milli_Seconds();
+        while (HAL_GPIO_Read(RI_UC) && !TIMEOUT(t0, 10000))
+        {
+            HAL_Delay_Milliseconds(1); // just wait
+        }
+        if(!HAL_GPIO_Read(RI_UC))
```

Comment:
We should set `ok = true` and `ok = false` depending on whether the RI_UC/V_INT indicates modem is actually powered off within the timeout or not. Probably worked in my functional tests since _pwr was still set to false...

Reply:
👍